src/hierarchies/nnw_hierarchy.cpp

```diff
@@ -12,84 +12,120 @@
 #include "../utils/proto_utils.hpp"
 #include "../utils/rng.hpp"
 
-void NNWHierarchy::check_hypers_validity() {
-  unsigned int dim = mu0.size();
-  assert(lambda0 > 0);
-  assert(dim == tau0.rows());
-  assert(nu0 > dim - 1);
-
-  // Check if tau0 is a square symmetric positive semidefinite matrix
-  assert(tau0.rows() == tau0.cols());
-  assert(tau0.isApprox(tau0.transpose()));
-  Eigen::LLT<Eigen::MatrixXd> llt(tau0);
-  assert(llt.info() != Eigen::NumericalIssue);
+//! \param prec_ Value to set to prec
+void NNWHierarchy::set_prec_and_utilities(const Eigen::MatrixXd &prec_) {
+  state.prec = prec_;
+
+  // Update prec utilities
+  prec_chol = Eigen::LLT<Eigen::MatrixXd>(prec_);
+  prec_chol_eval = prec_chol.matrixL().transpose();
+  Eigen::VectorXd diag = prec_chol_eval.diagonal();
+  prec_logdet = 2 * log(diag.array()).sum();
 }
 
-void NNWHierarchy::check_state_validity() {
-  // Check if tau is a square matrix
-  unsigned int dim = mean.size();
-  assert(dim == tau.rows());
-  assert(dim == tau.cols());
-  // Check if tau is symmetric positive semi definite
-  assert(tau.isApprox(tau.transpose()));
-  assert(tau_chol_factor.info() != Eigen::NumericalIssue);
+void NNWHierarchy::check_spd(const Eigen::MatrixXd &mat) {
```

Comment:
do we have any test for this? I think it's crucial as it will raise exceptions.
If not, I would adapt the check_pos_definite function in stan/math

https://github.com/stan-dev/math/blob/develop/stan/math/prim/err/check_pos_definite.hpp

since they have thoroughly tested it

Reply:
I didn't know it existed. In hindsight it was kind of obvious. Stan really has it all, hasn't it?